File: ideal-core/src/main/java/cn/idealframework2/idempotent/RedisTemplateIdempotentHandler.java
package cn.idealframework2.idempotent;

import cn.idealframework2.spring.RedisTemplateUtils;
import org.springframework.data.redis.core.StringRedisTemplate;

import javax.annotation.Nonnull;
import java.time.Duration;
import java.util.UUID;

/**
 * 基于RedisTemplate的幂等处理器
 *
 * @author 宋志宗 on 2022/10/13
 */
public class RedisTemplateIdempotentHandler implements IdempotentHandler {
  private final String lockValue = UUID.randomUUID().toString().replace("-", "");
  private final String prefix;
  private final Duration expire;
  private final StringRedisTemplate redisTemplate;

  public RedisTemplateIdempotentHandler(@Nonnull String prefix,
                                        @Nonnull Duration expire,
                                        @Nonnull StringRedisTemplate redisTemplate) {
    this.prefix = prefix;
    this.expire = expire;
    this.redisTemplate = redisTemplate;
  }

  @Override
  public boolean idempotent(@Nonnull String key) {
    String redisKey = prefix + ":" + key;
    return RedisTemplateUtils.tryLock(redisTemplate, redisKey, lockValue, expire);
  }

  @Override
  public void release(@Nonnull String key) {
    String redisKey = prefix + ":" + key;
    RedisTemplateUtils.unlock(redisTemplate, redisKey, lockValue);
  }
}

File: ideal-core/src/main/kotlin/cn/idealframework2/idempotent/coroutine/RedisTemplateIdempotentHandler.kt
package cn.idealframework2.idempotent.coroutine

import cn.idealframework2.spring.RedisTemplateUtils
import kotlinx.coroutines.reactor.awaitSingleOrNull
import org.springframework.data.redis.core.ReactiveStringRedisTemplate
import java.time.Duration
import java.util.*

/**
 * 基于RedisTemplate的幂等处理器
 *
 * @author 宋志宗 on 2022/10/13
 */
class RedisTemplateIdempotentHandler(
  private val prefix: String,
  private val expire: Duration,
  private val redisTemplate: ReactiveStringRedisTemplate
) : IdempotentHandler {
  private val lockValue = UUID.randomUUID().toString().replace("-", "")

  override suspend fun idempotent(key: String): Boolean {
    val redisKey = "$prefix:$key"
    return redisTemplate.opsForValue()
      .setIfAbsent(redisKey, lockValue, expire).awaitSingleOrNull() == true
  }

  override suspend fun release(key: String) {
    val redisKey = "$prefix:$key"
    RedisTemplateUtils.unlock(redisTemplate, redisKey, lockValue).awaitSingleOrNull()
  }
}
